Spellchecker CI

shall0pass opened the pull request:
I did very little editing.  Found https://github.com/marketplace/actions/check-spelling#basic-configuration as a possible spellchecker.  It could probably use a lot more optimizing for our needs, but I thought it was worth looking at.  I was able to get it up and running with ease, so it's not a lot of setup.

I edited the allow list and pasted in most of the first run words it caught.  Most are commands, from what I assume is the AQL pages, and github user names.  There were a few that I deleted out of the list because they just looked like misspellings without any context and should still show up as issues in the latest run on my branch.

Feel free to edit to your hearts content if it looks like this will work.

shall0pass commented:
There are 295 flagged words that need to be added to an ignore list or corrected.  I have all of the contributors in a single file.  There are additional libraries available that could eliminate some of the need for so many exclusions.  Bash/Java/Node/NPM are all available.  
The report is here: https://github.com/shall0pass/actual-docs/commit/4b0255ebb15a0e838f09f56c557899278c005355#commitcomment-120889190

shall0pass commented:
I've gotten it down to 77 words and a few pattern errors.  Some of the words may be ok, but some of them appear to be API commands where the camel case may not be correct.

https://github.com/shall0pass/actual-docs/commit/6b1d2960c68dbe9ce4c0a4b1df196b9dccf0887e#commitcomment-120928060

shall0pass commented:
Ok.  I think once the other spelling PR is merged, I'll merge master back into my branch, add the remaining terms to the allow list (as long as we don't see anything else that needs fixed) and it'll pass.

The latest report is here: https://github.com/shall0pass/actual-docs/actions/runs/5487604536
There are 70 words (some of which are already addressed in #229) that are throwing an error, and all pattern checks are passing.

I've fixed a few the double words (only only) and (the the).

shall0pass commented:
Ok.  It's down to 12, of which I **think** all but 1 are addressed in #229 .  

Is accs intended to be shorthand for accounts?

j-f1 commented:
> Is accs intended to be shorthand for accounts?

Looks like it, feel free to expand it.

shall0pass commented on .github/actions/spelling/allow/actualapi.txt:
```diff
@@ -0,0 +1,32 @@
+addTransactions
```
https://github.com/check-spelling/check-spelling/wiki/

It looks like it should.  I'll check later to see if I missed a configuration option.

shall0pass replied:
While I'm thinking about it.... There are multiple pages and links that don't use camelCase and get caught. Which prompted me to ignore words inside '',"", and ``.  If we could fix those links and pages to use camelCase, we could eliminate some of the regex patterns I added.

shall0pass commented:
I've removed the api exclusion file and it's all passing.  I also went ahead to fixed the few leftover spelling errors.
The files I modified the most are 

contributers.txt  -- these are the only ones that are caught because they don't have the @ symbol before them in the docs in a few locations.
keywords.txt
excludes.txt -- bottom 5

line_forbidden.patterns -- I commented out the into/in to rule

patterns.txt  -- I used the autogenerated patterns, which has a comment where they start and the bottom 6.

shall0pass commented on .github/actions/spelling/patterns.txt:
```diff
@@ -75,7 +75,7 @@
 
 # disable fields that are encapsulated in quotes. These are usually commands in how-to guides or section pages.
-'.+[']
+# '.+[']
```
Disabling this line finds one misspelling.

offbudget found on this page within the table.  https://actualbudget.org/docs/api/reference#accounts

shall0pass commented on .github/actions/spelling/patterns.txt:
```diff
@@ -75,8 +75,8 @@
-# '.+[']
-`.+[`]
+'.+[']
+# `.+[`]
```
Disabling backquotes finds one misspelling.

autocompletes -- only found in the release notes.  It's probably best to disable this change the word.

shall0pass commented on .github/actions/spelling/patterns.txt:
```diff
@@ -77,7 +77,7 @@
 '.+[']
 # `.+[`]
-".+["]
+# ".+["]
```
This didn't remove any errors.  Removing code.

j-f1 commented on blog/2023-03-13-release-23.3.2.md:
```diff
@@ -27,7 +27,7 @@ Version: 23.3.2
 - [#247](https://github.com/actualbudget/actual/pull/247) Route aggregate queries in transaction grouped mode through the correct layer to remove deleted transactions — thanks @jlongster
 - [#743](https://github.com/actualbudget/actual/pull/743) (nordigen) fallback to bookingDate if valueDate is not set — thanks @MatissJanis
-- [#742](https://github.com/actualbudget/actual/pull/742) (nordigen) check server status before linking accs — thanks @MatissJanis
+- [#742](https://github.com/actualbudget/actual/pull/742) (nordigen) check server status before linking  — thanks @MatissJanis
```
```suggestion
- [#742](https://github.com/actualbudget/actual/pull/742) (nordigen) check server status before linking accounts — thanks @MatissJanis
```

j-f1 commented on blog/2023-01-12-release-23.1.12.md:
```diff
@@ -9,7 +9,7 @@ hide_table_of_contents: false
 The release has notable of improvements of:
 
-- Read-only responsive view, this replaces our mobile apps, it is notable that this is read only only at this stage.
+- Read-only responsive view, this replaces our mobile apps, it is notable that this is read-only only at this stage.
```
Typo: “read only only”
```suggestion
- Read-only responsive view, this replaces our mobile apps, it is notable that this is read-only at this stage.
```

j-f1 commented on blog/2023-07-04-release-23.7.0.md:
```diff
@@ -139,7 +139,7 @@ Version: 23.7.0
 - [#1256](https://github.com/actualbudget/actual/pull/1256) Remove lingering references to `locationPtr` after `react-router` upgrade — thanks @j-f1
 - [#1257](https://github.com/actualbudget/actual/pull/1257) Moving more components from `common.tsx` to separate files inside the `common` folder — thanks @MatissJanis
 - [#1258](https://github.com/actualbudget/actual/pull/1258) Reorganized accounts directory. Pulled our Header functions to make the accounts.js smaller and more manageable. — thanks @carkom
-- [#1259](https://github.com/actualbudget/actual/pull/1259) Refactoring some usages of legacy `Select` component to autocompletes or `CustomSelect` — thanks @MatissJanis
```
This is a legitimate word imo

shall0pass replied:
Ok.  I'll put autocomplete in the keywords file to allow it.  It didn't sound right when I read it, but I definitely agree it should be a legitimate word.

shall0pass commented on .github/actions/spelling/patterns.txt:
```diff
@@ -74,7 +74,7 @@
 [#][^$\W]*
 
-[..\/].+\w
+# [..\/].+\w
```
I don't think this regex pattern is written correctly, but these are the errors when disabled.

List of errors:
abcd -- restore.md -- 1. Unzip the backup budget `.zip` file. The filename should look like: `yyyy-mm-dd-My-Finances-abcd1234.zip`

asdf -- local.md -- Consider using a tool like [`nvm`](https://github.com/nvm-sh/nvm) or [`asdf`](https://asdf-vm.com) to install and manage multiple versions of Node.js.

Blix -- index.md -- All strings are matched case-insensitive. An `imported payee` of "PuBlix" will match a condition that is "contains 'publix'".

Crd -- payees.md -- Usually this ends up with some ugly names like `Target Debit Crd Ach Tran Co Id:Xxxxx15170`.

creditcards -- docs-sidebar.js -- 'budgeting/creditcards',

heidi -- modify-transfers.md -- 3. Open the `db.sqlite` file with your preferred tool [sqlite3 cli](https://www.sqlite.org/cli.html), [heidiSQL](https://www.heidisql.com/), ect

jointaccounts
keyout
managefiles
returnsandreimbursements
Rezip
tipstricks
transations
verison
Xxxxx

shall0pass replied:
creditcards, jointaccounts, managefiles, returnsandreimbursements and tipstricks are all pages and could be renamed with camelCase to fix the error.

shall0pass commented on .github/actions/spelling/patterns.txt:
```diff
@@ -72,7 +72,7 @@
 
 # docusaurus links that begin with '#'
-[#][^$\W]*
+# [#][^$\W]*
```
Disabling this exposed these:  Again, they could all pass if we use camelCase.

addtransactions
categorygroup
deleteaccount
importtransactions
payeerule
undoredo

shall0pass commented:
These are the last few.

abcd  -- Part of a sample file name.  Ok.
Blix -- Part of a sample Import Transaction.  Ok.
Crd -- Part of a sample Import Transaction.  Ok.
keyout -- Option in openssl command.  Ok.
Xxxxx -- Part of a sample Import Transaction.  Ok.

I'm tempted to make carve outs for these in the acceptable words list so they pass.  Thoughts?

This PR is becoming much larger than I anticipated, especially with the file renaming to camelCase.  The APIList.js file caught me by surprise.  If there is a better way to handle those links, let me know.  If I need to split that into a separate PR, I can.  I think I've fixed all of the links that go with it, but I sure would appreciate a double check.

j-f1 commented on .github/actions/spelling/candidate.patterns:
Does this come from somewhere? Is there a way we can avoid having to manage a 500+ line file ourselves?

shall0pass replied:
This file is part of the default project files. (https://github.com/check-spelling/spell-check-this)  I didn't edit this file from the default state.  We could try deleting or renaming it so it doesn't get processed to see what the results are.

j-f1 replied:
That’s fine to keep it then, can you add a comment linking to that original file so that we know where to get updates in the future if needed?

shall0pass replied:
Oh whoops!  I just deleted it.  There weren't any new errors.

j-f1 replied:
If it isn’t used during the check then I think it should be removed.

j-f1 commented on docs-sidebar.js:
```diff
@@ -122,9 +122,9 @@ const sidebars = {
           ],
         },
-        'getting-started/tipstricks',
+        'getting-started/tipsTricks',
```
This may break existing links because URLs are sorta case sensitive. Would you be willing to rename these to kebab case (`getting-started/tips-tricks`) and add corresponding redirects to the `_redirects` file? It would be nice to avoid capital letters in the URLs if possible.

shall0pass replied:
I think I've gotten all of the redirects in.

j-f1 commented on docs/api/APIList.js:
```diff
@@ -7,7 +7,7 @@ export default function APIList({ title, sections }) {
       <ul>
         {sections.map(name => {
-          let id = name.replace(/[ -]/g, '-').toLowerCase();
+          let id = name.replace(/[ -]/g, '-');
```
This will also break any existing links to sections, I think.

shall0pass replied:
Are the existing links you're referring to inside this repository, or from the classic actualbudget.com?

shall0pass replied:
Could we do this?

          let id = name.replace(/[A-Z]/g, '-$&'); 
          id = id.replace(/[ -]/g, '-').toLowerCase();
          id = id.replace(/^-/g, '');

It adds dashes between camelCase words in addition to changing spaces to dashes, and then finally removing a dash at the beginning of the string if it exists.

j-f1 replied:
Existing links to actualbudget.org from elswhere (discord, links in current/old versions of the app, etc)

shall0pass replied:
Hmm. Ok.  We'll have to add exclusions for the links then.  The code will create #deletepayee, which by itself won't create a misspelling but anytime there is a link created to it, it will.

j-f1 replied:
Maybe add an `<a name="deletepayee">` tag with an ignore comment right above?

It would be awesome if there was a way to get the camel case headings to be auto generated but it doesn’t appear to be easy so it’s ok to keep the custom heading IDs in place.

j-f1 commented on docs/api/reference.md:
```diff
@@ -442,31 +442,31 @@ Delete a payee.
+Update fields of a payee rule. `fields` can specify any field described in [`PayeeRule`](#Payee-rule).
 
-#### `deletePayeeRule`
+#### `deletePayeeRule` {#deletePayeeRule}
```
Can these not be inferred?

shall0pass replied:
The inferred link is #deletepayeerule.  Because it's not camelCase and doesn't have dashes, it gets caught. If there's a way to have default links use a dash when camelCase is found, that would be awesome but I haven't found it in the documentation yet.

shall0pass replied:
If we delete the backquotes on these sections, the link will be inferred.  But we would also need to remove the .toLowerCase() operation in the API page for it to match.

Let me know how you want to proceed.  I don't know how many external links really exist to these individual sections of the documentation.  Personally, I think the page looks better without the backquotes from a formatting perspective.
This shows deleteAccount changed vs the current formatting.  You can also see the link is inferred as camelCase here.

![image](https://github.com/actualbudget/docs/assets/20625555/201891c8-abe3-43f5-b6cf-3b8faa2b8de2)

j-f1 replied:
I don’t have any strong opinions here, feel free to do whatever you think is best!

shall0pass replied:
I think I'll leave it as-is and keep the 5-ish keyword links in the allowed file.  If anyone ever adds more to these sections and it becomes confusing to keep them updated, we can re-evaluate then.  As of now, these pages are fairly static with no content changes.

j-f1 commented on docs/getting-started/tipsTricks.md:
```diff
@@ -3,7 +3,7 @@ sidebar_position: 1
 ---
 
-## Undo/redo
+## Undo/redo {#undoRedo}
```
```suggestion
## Undo/redo {#undo-redo}
```

shall0pass replied:
done

j-f1 commented on docs/getting-started/tipsTricks.md:
```diff
@@ -31,7 +31,7 @@ A few global shortcuts:
 - <Key k="f" /> Show only selected transactions.
 - <Key k="d" /> Delete selected transactions.
-- <Key k="a" /> Set account for selected transations.
+- <Key k="a" /> Set account for selected transactions.
```
Nice!

j-f1 commented on .github/actions/spelling/allow/keywords.txt:
```diff
@@ -0,0 +1,87 @@
+authelia
+autocompletes
+Blix
+caddyserver
```
This only appears in a URL, doesn’t the existing URL exclusion avoid needing this?

shall0pass replied:
It may have been entered before I had some of the other patterns working correctly.  I should probably delete the whole file again and only copy in what is flagged.  It should have been caught with the URL regex.

shall0pass replied:
I've rerun the keywords.  There were about 10 more that fell of the list, so they may have been in URLs also.

j-f1 commented on static/_redirects:
```diff
@@ -28,6 +28,9 @@
 # /docs/budgeting/categories/       /docs/budgeting/categories/
+/docs/budgeting/returnsandreimbursements    /docs/budgeting/returns-and-reimbursements
+/docs/budgeting/budgeting/creditcards       /docs/budgeting/budgeting/credit-cards
+/docs/budgeting/budgeting/jointaccounts     /docs/budgeting/budgeting/joint-accounts
```
Should `budgeting` be in here twice? Also, can you move this section to the bottom and comment that these files were renamed? The current redirects are all going from the .com docs to the current docs.

shall0pass replied:
Agh.  No it shouldn't be.  Must have been copy/paste happy.

j-f1 reviewed (commented):
One more comment, otherwise LGTM — let me know when you’re ready to merge!

j-f1 reviewed (approved):
Thank you for getting this working!

shall0pass commented:
Absolutely.  It was a fun learning experience.
